### DISCUSSION TITLE
Learning how to syntax highlight

### DISCUSSION THREAD

**Discussion by CodingPupper3033** [intellij-dlanguage/intellij-dlanguage#648] [General]
Hey! I know this isn't the preferred way to contact you, however, I couldn't find another way to message you. How did you learn how to make something like this. I am looking at Intellij documentation and it made no sense to me. It would continuously bring up topics that you are supposed to understand, however, I couldn't find any documentation on it.

What I understand / could do:
 - https://plugins.jetbrains.com/docs/intellij/prerequisites.html
 - https://plugins.jetbrains.com/docs/intellij/language-and-filetype.html
 - AST Trees (I think anyway, I read the wiki and it wasn't too difficult to understand)

What was mindboggling:
 - https://plugins.jetbrains.com/docs/intellij/grammar-and-parser.html
 - https://plugins.jetbrains.com/docs/intellij/implementing-parser-and-psi.html

Some specific questions:
 - Where do I find less confusing documentation?
 - What is a Token Type and how is that different from an Element Type
 - What does the bnf file do?
 - What does the bnf file say?
 - Is there some general knowledge I should know/read about from somewhere other than Intellij's website?
 - Any good video tutorials?

My Goal:
 - I am taking a data structures class and I am learning graphs. I am trying to make my own UTF-8 file extension (.grph) I have each line declaring a node. If the line is a comment it starts with "--". To separate nodes and connections I use ":". To separate a connection and its weight I use a ".". Example line "E: D.5, G.1, H.1"
 - I made a parser (using Kotlin input streams) to convert this file to a useable graph, but I think it would be cool if I could see the file highlighted, seeing comments vs nodes vs edges vs weights. It isn't strictly necessary, but I think it would be cool.

TLDR:
 - I have no clue how you learned to make something like what you have created. I was hoping you could help me or preferably send me to resources that could let me wrap my head around what you have done. 

Sorry to waste your time, I just can't find any documentation that I could understand.

**Comment by rikkimax** [comment]
So a [BNF ](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form) file is a specification for a language grammar. It is typically used in parser generators like what is recommended to be used with IntelliJ IDEA.

A good general book that is highly recommended in the D community as a starting point on compiler development is ["the dragon book"](https://en.wikipedia.org/wiki/Compilers:_Principles,_Techniques,_and_Tools).

On a related note, [Graphviz Dot ](https://graphviz.org/doc/info/lang.html) may be of interest to you as it is a graphing tool that covers graphs.

**Comment by CodingPupper3033** [comment]
So, how do I make this bnf file for the Intellij plugin? I can see the example requires Token Types and Element Types, how do I make these correctly and what do they do?

Also, thanks so much for responding, I appreciate your time!


Edit: I forgot to mention, thanks for the Graphviz Dot link, looks helpful!

**Comment by rikkimax** [comment]
So what you are looking for is to do with PSI.

But its probably best to start with: https://plugins.jetbrains.com/docs/intellij/welcome.html

**Reply by CodingPupper3033** [reply]
I was looking over that website and that is where I got stuck. I looked over the PSI section too, and it was very confusing for me

**Comment by SingingBush** [comment]
When I started getting involved with this project there was very little documentation, it's certainly better now than it used to be. Jetbrains have got a lot better at sharing the dark art of Intellij plugin development. Although I've had a go at working with the grammar/parser side of things my main focus has always been trying to simply keep the plugin working (the Intellij API changes far more frequently now than it used to) and gradually improve integration with tools like dub. Contributions toward improving grammar/parser code in this projet are greatly needed.

**Comment by CodingPupper3033** [comment]
Thank you guys again for helping!
